src/python/pants/scm/subsystems/changed.py

```diff
@@ -38,7 +52,36 @@ def register_options(cls, register):
              help='Calculate changes since this tree-ish/scm ref (defaults to current HEAD/tip).')
     register('--diffspec',
              help='Calculate changes contained within given scm spec (commit range/sha/ref/etc).')
-    register('--include-dependees', choices=['none', 'direct', 'transitive'], default='none',
+    register('--include-dependees', type=IncludeDependees, default=IncludeDependees.none,
              help='Include direct or transitive dependees of changed targets.')
     register('--fast', type=bool,
              help='Stop searching for owners once a source is mapped to at least one owning target.')
+
+
+class ScmWrapper(datatype([('scm', SubclassesOf(Scm))])): pass
```

Comment:
As a cute hack, I believe that if you remove the implementation of equality for this type (which would require not extending `datatype`, since we guard against that), and allow it to default to `id` (ie, no two `ScmWrappers` are equal unless they have the same memory address), then caching of these rules will work out, because any subgraph that transitively depends on the `ScmWrapper` will re-run for each call, while other subgraphs that don't consume the parameter will be unaffected.

Commented on #6598 in this vein.

Reply:
Is this a cute hack? This looks extremely legitimate, especially if the uncacheable `Console` can be used alone to make an `@console_rule`.

Reply:
Sorry, yes: very legitimate. I only refer to it as a hack due to the need to formalize it a bit more as mentioned in the comment on #6598.

Reply:
Definitely sounds like a hack to me... I can imagine debugging an issue around here in the future, and it being pretty non-obvious that this rule should be re-running because the type which happens to implement a function it calls has its equality overridden, and eventually having an "a-ha" moment after a lot of debugging... (And ideally `--visualize` and friends would show something as uncacheable). A very cute hack, but definitely not what I'd want long-term without a lot more formalisation...

Reply:
As a quick, obvious way to avoid backsliding, one approach would be to add a tiny wrapper type similar to https://github.com/pantsbuild/pants/blob/b1703795f864bd5e9f981d6b404967d6c83f3ca9/src/python/pants/engine/objects.py#L157-L166 that would carefully explain itself and ban overriding equals.

You'd then create uncachable types via `ScmWrapper = Uncachable.of(Scm)`.

Reply:
I think `Uncachable.of(Scm)` is a fantastic idea!